recursion/writing-recursive-code.md

```diff
@@ -0,0 +1,183 @@
+
+| Function Call | Argument(s) | Return value |
+| ------------- | ----------- | ------------ |
+
+## Walk-through: Summing Natural Numbers
+
+In mathematics, _natural numbers_ are numbers used for counting and ordering. For this situation, we can define natural numbers as all positive integers starting at 1.
+
+Let's consider this programming problem:
+
+Create a function named `sum_natural_numbers` that is responsible for summing all natural numbers up to and including the input, `num`. We can assume that `num` is always a positive integer. This function should return the sum.
+
+Examples:
+
+| `num` | Example           | Sum  |
+| ----- | ----------------- | ---- |
+| `3`   | 3 + 2 + 1         | `6`  |
+| `4`   | 4 + 3 + 2 + 1     | `10` |
+| `5`   | 5 + 4 + 3 + 2 + 1 | `15` |
+| `1`   | 1                 | `1`  |
+
+### Determine the Base Case
+
+The problem states that natural numbers begin at 1.
+
+We can determine the return value of `sum_natural_numbers(1)` without using recursion! This algorithm needs to stop when `num` is `1`.
+
+Therefore, our base case is when `num == 1`.
+
+### Describe the Repeated Smaller Problem
+
+The problem states that the function "computes the sum of all natural number up to and including `num`."
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: 7lIVkd
+* title: Writing Recursive Code
+##### !question
+
+Consider the `sum_natural_numbers` problem.
+
+What is one way we describe the repeated problem? What is the problem that gets divided into smaller versions of itself?
+
+##### !end-question
+##### !placeholder
+
+The repeated problem for sum_natural_numbers is...
+
+##### !end-placeholder
+##### !explanation
+
+We might describe the repeated problem like so:
+
+The repeated problem is addition. We need to add the sum of all natural numbers calculated so far and a new number.
+
+##### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### Determine the Information the Smaller Problem Needs
+
+Each smaller version of the problem needs an input number, `num`.
+
+Each recursive call should bring the number closer to the base case, `1`. Because `1` is the smallest natural number, and all other integers are greater than `1`, we should start our recursion from a larger integer and decrement `num`.
+
+By decrementing, we can predict that `num` will eventually become `1` and end the recursion.
+
+### Determine the Information We Want Back
+
+The value we need to `return` from each call is the sum of all natural numbers up to and including `num`.
+
+### Create Example Tables
+
+The following table represents the system stack when calling `sum_natural_numbers(4)`. The top of the stack is the first row, `sum_natural_numbers(1)`. This function call is at the top of the stack because it is the base case, so it is the final function call. With stacks, the last item pushed in is the first item popped out (LIFO!).
```

Comment:
system stack → call stack
```suggestion
The following table represents the call stack when calling `sum_natural_numbers(4)`. The top of the stack is the first row, `sum_natural_numbers(1)`. This function call is at the top of the stack because it is the base case, so it is the final function call. With stacks, the last item pushed in is the first item popped out (LIFO!).
```